docs/proposals/mapper-design.md

```diff
@@ -0,0 +1,117 @@
+We can define an event as a group of read/write operations on a device. A device may support multiple such events. Each of this events should be supplied through config-file to event manager.
+An Operation can be defined using below defined structure
+```go
+// Operation is structure to define device operation
+type Operation struct {
+	// Action can be one of read/write corresponding to get/set respectively
+	Action string
+	// Register is physical register where operation needs to be performed
+	Register string
+	// Value is the value to be written in case of write action
+	Value string
+}
+```
+An Event is a set of operation(s) and can be defined using below structure
+```go
+// Event is structure to define a device event
+type Event struct {
+	// Name is the name of the event
+	Name string
+	// Operations is the list of operations to be performed for performing this event
+	Operations []Operation
+}
+```
+**2) Scheduler**: Scheduler is used to perform event(s) after defined intervals. Schedule can be defined using below structure
+```go
+// Schedule is structure to define a schedule
+type Schedule struct{
+	// Name is name of the schedule. It should be unique so that a stop-chan
+	// can be made corresponding to name to stop the schedule.
+	Name string
+	// Frequency is the time in milliseconds after which this events are to be performed
+	Frequency int 
+	// Events is list of events to be performed in this schedule
+	Events []Event
+}
+```
+
+**3) Watcher**: Watcher has 3 responsibilities: 
+
+a) To scan devices(wireless)/wait for device to turn on(wired) and connect to the correct device once it is Online/In-Range. It can use MAC address or any unique address provided by devices. In case of wired devices, GPIO can be an option.
+
+b) Keep a watch on the expected state of the twin-attributes of the device and perform the event(s) to make actual state equal to expected.
+
+c) To report the actual state of twin attributes.
+
+**4) Data-Converter**: Data received from the devices can be in complex formats. eg: HexDecimal with bytes shuffled. This data cannot be directly understood by KubeEdge.
+The responsibility of data-converter is the convert the readings into a format understood by KubeEdge.
+Many protocols have a standard defined for the reading returned by the devicce. Hence a common/configurable logic can be used. 
+
+**5) Health-Checker**: Health-Checker can be used to periodically report the state of the device to KubeEdge.
+This can be an optional component as not all devices support health-checking. In-future can be extended to report battery-state, malfunctioning when kubeedge supports these attributes.
+
+**6) Controller**: Controller should expose API's for CRUD operations for managing Events, Schedules, Watchers, Data-Converters, Health-Checkers.
```

Comment:
we need to be more specific on the APIs. REST/gRPC/?. The apis will need documentation describing the input/ouputs etc. maybe we can add more details . thoughts ?

Reply:
I initially plan to have mqtt based API's like the one's in device-twin. Maybe I will come-up with a API-Documentation.

Reply:
Here are two mode to call API. async/sync mode. In sync mode, we  use service bus to pass messages. and in async mode we use eventbus to pass messages.

Reply:
How can we access the controller from cloud？

Reply:
@zhenghaibo1008 , currently KubeEdge doesn't have a cloud part, so in v0.1 we only plan to allow access to the controller from the edge via MQTT

Reply:
How do you wan't to deploy mapper? modules or docker images?

Reply:
Currently we want to deploy mappers as a process, After cloud-part in there we will support both docker and process.

Reply:
If we deploy mapper as docker, we can use configmap